Fix deserialization of events with entitlements and add better error logging

lealobanov opened the pull request:
Resolves issue #66 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **New Features**
	- Introduced a new example for debugging transaction fetching by ID.
	- Added a new JSON structure for event capabilities, enhancing metadata representation.
	- Added a new test for account capability event encoding and decoding.

- **Bug Fixes**
	- Improved error handling and logging for event deserialization and transaction results.

- **Chores**
	- Updated `.gitignore` to include `myenv` for better environment management.
	- Renamed `AuthAccountKind` to `AccountKind` for clarity in naming conventions.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

lealobanov committed:
Debugging events decoding

lealobanov committed:
Clean up logs

lealobanov committed:
Clean up

lealobanov commented on myenv/bin/activate.csh:
```diff
@@ -0,0 +1,26 @@
+# Unset irrelevant variables.
+deactivate nondestructive
+
+setenv VIRTUAL_ENV "/Users/lealobanov/Desktop/Projects/flow-py-sdk/myenv"
```
Updated gitignore, will delete `myenv` before merging

lealobanov committed:
Clean up logging

lealobanov commented on examples/transactions_examples.py:
```diff
@@ -281,3 +281,22 @@ async def run(self, ctx: Config):
             self.log.info(f"transaction script: {transaction.script.decode('utf-8')}")
+
+
+class DebuggingFetchTransactionByIdExample(Example):
```
Added example which makes a mainnet call to the failing tx ID `ef6e6dcba87c5f853e0ec940dd382d609ffe11109c28ed239843d71618fa1d71`, will remove this before merging changes

lealobanov commented on failed_events.json:
This is an output of the failing event payloads for reference, will remove before merging changes

lealobanov committed:
Delete myenv directory

lealobanov commented on flow_py_sdk/client/entities.py:
```diff
@@ -287,13 +312,25 @@
+            except Exception as e:
+                print(f"Failed to deserialize event {i}: {e}")
+                continue
+    
```
Updated

lealobanov committed:
Lint

lealobanov committed:
Merge branch 'bug-66' of https://github.com/janezpodhostnik/flow-py-sdk into bug-66

janezpodhostnik commented on flow_py_sdk/client/entities.py:
```diff
@@ -1,4 +1,5 @@
 import json
+import logging
```
This Is a good idea ^

janezpodhostnik reviewed (commented):
Thank you very much for the contribution!

Can you add a short explanation of why the events failed to de-serialize, I'm not quite sure I get it. Was there an unexpected dict/list in the event? 

As for the error handling: 
- I very much like the Idea of extra logging, that will make it a lot easier to figure out whats wrong next time.
- I'm not sure I like the idea of just logging and continuing. That seems very dangerous. I would rather the error was logged and the exception re-thrown. There could be an optional parameter where instead of re-throwing it would just return the json object if decoding failed. What do you think?

lealobanov commented on flow_py_sdk/client/entities.py:
```diff
@@ -288,12 +308,22 @@ def from_proto(
         id: bytes,
     ) -> "TransactionResultResponse":
+        events = []
+        for i, event_proto in enumerate(proto.events):
+            try:
+                event = Event.from_proto(event_proto)
+                if event is not None:
+                    events.append(event)
+            except Exception as e:
+                logging.error(f"Failed to deserialize event {i}: {e}")
+                continue
+
         return TransactionResultResponse(
             id_=id,
             status=proto.status,
             status_code=proto.status_code,
             error_message=proto.error_message,
-            events=[Event.from_proto(e) for e in proto.events],
+            events=events,
```
@janezpodhostnik initially when debugging I considered introducing a max limit on number events processed per transaction as I thought it might be causing the issue... in the end it didn't seem to have much impact on this specific transaction so I opted not to pursue these changes for now.

lealobanov replied:
Will resolve this comment for time being and we can look into introducing a restriction on event count if we run into further issues

lealobanov committed:
Raise exception instead of skipping decoding event when failure

lealobanov commented:
Hi @janezpodhostnik ,

- Just updated the exception handling to raise an exception and exit the decoding if one of the events failed to decode (instead of continuing).

I think the source of the bug might have been the recent migration to Cadence 1.0, as I noticed all of the events which failed to decode had complex nested dict/list fields such as `authorizations` and `staticType` found in the `EntitlementConjunctionSet` class.  Entitlements seem to be a new syntax introduced since the C1.0 release (https://www.youtube.com/watch?v=aZ1aCBtPpXo), so I suspect this might be why this wasn't throwing an error before.

Thanks!

franklywatson commented on flow_py_sdk/cadence/decode.py:
```diff
@@ -17,32 +17,53 @@ def add_cadence_kind_decoder(t: Type[Kind]):
+                decoder = _cadence_decoders[type_]
+                return decoder(obj)
+
+        # Recursively handle nested structures
+        if isinstance(obj, dict):
+            for key, value in obj.items():
+                if isinstance(value, dict):
+                    obj[key] = decode(value)  # Recursive call for nested dict
+                elif isinstance(value, list):
+                    obj[key] = [
+                        decode(item) for item in value
+                    ]  # Recursive list handling
+
+        elif isinstance(obj, list):
+            obj = [
+                decode(item) for item in obj
+            ]  # Handle obj itself as a list if list type
+
```
@lealobanov what do we think about this suggestion? Given decode is a heavily used method I would err on making ti as robust as we can

janezpodhostnik committed:
fix

janezpodhostnik commented on flow_py_sdk/client/entities.py:
```diff
@@ -149,17 +150,36 @@ def __init__(
+            logging.error(
+                f"JSON decode error for event {event_index} with payload: {payload[:100]}... Error: {str(e)}"
+            )
+            self.value = None
```
I like the addition of logging, but I would not return None, as it can lead to confusing results for users (everything seems ok, but no value is returned). I would instead just re-raise the exception here. n

same goes for all try-catch blocks in this file: `flow_py_sdk/client/entities.py`.

janezpodhostnik commented:
> Just updated the exception handling to raise an exception and exit the decoding if one of the events failed to decode (instead of continuing).

Looks great, thank you. Can you do the same for `flow_py_sdk/client/entities.py`

I did some [digging](https://github.com/janezpodhostnik/flow-py-sdk/pull/68). And I realized that the recursive decoding should not be needed, since the cadence [json api objects](https://cadence-lang.org/docs/json-cadence-spec) already define if something can be a list or not, and the `EntitlementsKind` do correctly accept a list.

```py
class EntitlementsKind(Kind, ABC):
    def __init__(self, entitlements: list[EntitlementBaseKind]) -> None:
```

The error was that an element that should have been decoded was not due to the incorrect naming of "Account" (it was still "AuthAccount" since I forgot to change it from before cadence 1.0)

Changing that fixed the issue without the need of adding the recursive decoding.


If you want to pull those changes in your PR and cleanup the temporary things we can get this merged.

As for this:

> Some other preventative measures we could take on the decoding of events array:
> 
>     Imposing a max number of events that get decoded per transaction
>     Imposing a max limit on event payload size

That is a great idea, and the contribution would be much appreciated, but lets not add it to this PR to keep it smaller and because its supposed to be a bugfix.

lealobanov committed:
Update logging and exception handling

lealobanov committed:
Merge branch 'bug-66-test' of https://github.com/janezpodhostnik/flow-py-sdk into bug-66

lealobanov committed:
Lint

lealobanov committed:
Re-enable test

lealobanov committed:
Clean up

lealobanov committed:
Clean up

lealobanov commented:
Hi @janezpodhostnik ,

Good catch, I didn't notice that the naming of the field had changed - thank you so much for looking into it. I've just pulled in your branch and cleaned up the test code. I've run the latest changes against tx ID `ef6e6dcba87c5f853e0ec940dd382d609ffe11109c28ed239843d71618fa1d71` and can confirm it decoded successfully, so have gone ahead and removed this from the transaction_examples files.

lealobanov committed:
Clean up

lealobanov committed:
Update logging messages

lealobanov commented on flow_py_sdk/cadence/decode.py:
```diff
@@ -17,32 +17,53 @@ def add_cadence_kind_decoder(t: Type[Kind]):
+
+    except NotImplementedError:
+        logging.error(f"Decoding not implemented for object: {obj}")
+        raise
```
Incorporated these improved messages but left the status code returned as error

lealobanov commented on flow_py_sdk/cadence/__init__.py:
```diff
@@ -90,7 +90,7 @@
     PublicPathKind,
     PrivatePathKind,
-    AuthAccountKind,
+    AccountKind,
```
Can confirm simple_kinds.py has already been updated with these changes

lealobanov commented on flow_py_sdk/client/entities.py:
```diff
@@ -149,17 +150,36 @@
         self.payload: bytes = payload
-        self.value: cadence.Event = json.loads(payload, object_hook=cadence_object_hook)
+
+        try:
+            # Attempt to decode the payload
+            self.value: cadence.Event = json.loads(
+                payload, object_hook=cadence_object_hook
+            )
+        except json.JSONDecodeError as e:
+            logging.error(
+                f"JSON decode error for event {event_index} with payload: {payload[:100]}... Error: {str(e)}"
+            )
+            raise
+        except Exception as e:
+            logging.error(
+                f"Unexpected error deserializing payload for event {event_index} with payload: {payload[:100]}... Error: {str(e)}"
+            )
+            raise
```
Will hold off on these improvements in a separate PR

lealobanov committed:
Update logging messages

lealobanov committed:
Lint

janezpodhostnik reviewed (approved):
Looks great! Thank you for the contribution.
